src/js/index.js

```diff
@@ -0,0 +1,309 @@
+ event when the screenshot is complete.  The returned value (`{arguments}.0`) represents the path to the generated file.
+
+ In addition to the functions provided by Nightmare, we provide:
+
+    1. `sendKey(int keyCode)`: a function to simulate a keyboard event.  Fires `onKeySent` when complete.
+    2. `getFocusedElement()`: a function to get the object that is currently focused
+    3. `isFocused(selector)`: a function that returns true if the specified selector is currently focused.
+
+  See the tests in this package for examples of how this can all be used together.
+
+ */
+"use strict";
+var fluid = require("infusion");
+var gpii = fluid.registerNamespace("gpii");
+
+fluid.registerNamespace("gpii.tests.browser");
+
+gpii.tests.browser.loadTestingSupport = function () {
+    require("../../tests/js/lib/fixtures");
+    require("../../tests/js/lib/sanity");
+};
+
+var Nightmare = require("nightmare");
+var os        = require("os");
+var path      = require("path");
+
+gpii.tests.browser.init = function (that) {
+    that.nightmare = new Nightmare(that.options.nightmareOptions); // jshint ignore: line
+
+    fluid.each(that.options.eventBindings, function (rawEventOrEvents, fluidEvent) {
+        var rawEvents = fluid.makeArray(rawEventOrEvents);
+        fluid.each(rawEvents, function (rawEvent) {
+            that.nightmare.on(rawEvent, function () {
+                that.events[fluidEvent].fire(fluid.makeArray(arguments));
+            });
+        });
+    });
+
+    that.events.onReady.fire(that);
+};
+
+gpii.tests.browser.execute = function (that, eventName, fnName, args) {
+    var argsArray = fluid.makeArray(args);
+    that.nightmare[fnName].apply(that.nightmare, argsArray).run(function (err, result) {
+        if (err) {
+            that.events.onError.fire(err);
+        } else {
+            that.events[eventName].fire(result);
+        }
+    });
+};
+
+// TODO:  Review and see why the process must be killed.
+gpii.tests.browser.end = function (that) {
+    that.nightmare.end();
+
+    // For some reason the "end" callback does not always execute, so we use an interval to take care of business
+    var millis = 0, intervalTimer = setInterval(function () {
+        millis += that.options.endInterval;
+        if (that.nightmare.ended === true) {
+            clearInterval(intervalTimer);
+            that.events.onEndComplete.fire(that);
+        } else if (millis > that.options.endTimeout) {
+            clearInterval(intervalTimer);
+
+            process.kill(that.nightmare.proc.pid);
+
+            fluid.log("Nightmare browser timed out and was killed manually....");
+            that.events.onEndComplete.fire(that);
+        }
+    }, that.options.endInterval);
+};
+
+
+gpii.tests.browser.executeAndReturnValue = function (that, eventName, fnName, args) {
+    var argsArray = args ? fluid.makeArray(args) : [];
+    that.nightmare[fnName].apply(that.nightmare, argsArray).run(function (error, result) {
```

Comment:
Take a look, I tidied things up considerably and removed one of the redundant execute functions.

I chose to keep the screenshot and pdf functions as convenience wrappers, as the test sequences would be a bit bloated otherwise.  All variations use common utility functions, including a new function to generate the "handler" function that will deal with Nightmare's response (or lack thereof).